File: runtimes/node/src/ratelimit.ts
export class TokenBucketRateLimiter {
    private capacity: number;
    private refillRate: number; // tokens per millisecond
    private tokens: number;
    private lastRefill: number;

    constructor(capacity: number, refillRatePerSec: number) {
        this.capacity = capacity;
        this.refillRate = refillRatePerSec / 1000;
        this.tokens = capacity;
        this.lastRefill = Date.now();
    }

    public allow(): boolean {
        const now = Date.now();
        const elapsed = now - this.lastRefill;
        this.lastRefill = now;

        this.tokens = Math.min(this.capacity, this.tokens + elapsed * this.refillRate);

        if (this.tokens >= 1) {
            this.tokens -= 1;
            return true;
        }
        return false;
    }
}

import { Redis } from 'ioredis';

const LUA_TOKEN_BUCKET = `
local key = KEYS[1]
local rate = tonumber(ARGV[1])
local burst = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local tokens_key = key .. ":tokens"
local timestamp_key = key .. ":ts"

local last_tokens = tonumber(redis.call("get", tokens_key))
if last_tokens == nil or last_tokens == false then
    last_tokens = burst
end

local last_refreshed = tonumber(redis.call("get", timestamp_key))
if last_refreshed == nil or last_refreshed == false then
    last_refreshed = 0
end

local delta = math.max(0, now - last_refreshed)
local filled_tokens = math.min(burst, last_tokens + (delta * rate))
local allowed = filled_tokens >= 1
local new_tokens = filled_tokens

if allowed then
    new_tokens = filled_tokens - 1
end

redis.call("setex", tokens_key, math.ceil(burst / rate), new_tokens)
redis.call("setex", timestamp_key, math.ceil(burst / rate), now)

if allowed then
    return { new_tokens, 1 }
else
    return { new_tokens, 0 }
end
`;

export class RedisRateLimiter {
    private redis: Redis;
    private rps: number;
    private burst: number;

    constructor(redis: Redis, rps: number, burst?: number) {
        this.redis = redis;
        this.rps = rps;
        this.burst = burst ?? Math.max(1, Math.floor(rps));
        
        // Register the script
        this.redis.defineCommand('allowRateLimit', {
            numberOfKeys: 1,
            lua: LUA_TOKEN_BUCKET,
        });
    }

    public async allow(key: string): Promise<[number, boolean]> {
        const now = Date.now() / 1000.0;
        // @ts-ignore - custom command
        const res = await this.redis.allowRateLimit(`ratelimit:${key}`, this.rps, this.burst, now);
        return [Number(res[0]), Boolean(res[1])];
    }
}

File: runtimes/rust/src/ratelimit.rs
use redis::Client;
use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
use std::sync::{Arc, RwLock};
use std::time::{SystemTime, UNIX_EPOCH};

const NANO_TOKEN: i64 = 1_000_000_000;

#[derive(Debug)]
pub struct ClientBucket {
    tokens: AtomicI64,    // stored as nano-tokens
    last_seen: AtomicU64, // Unix nanoseconds
    capacity: i64,        // nano-tokens
    refill_ns: i64,       // nano-tokens/ns (= tokens/s at nano scale)
}

impl ClientBucket {
    pub fn new(rps: f64, burst: i64) -> Self {
        let cap = (burst as f64 * NANO_TOKEN as f64) as i64;
        let now_ns = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_nanos() as u64;

        Self {
            tokens: AtomicI64::new(cap),
            last_seen: AtomicU64::new(now_ns),
            capacity: cap,
            refill_ns: rps as i64,
        }
    }

    pub fn allow(&self) -> (i64, bool) {
        let now_ns = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_nanos() as u64;

        let last = self.last_seen.swap(now_ns, Ordering::SeqCst);
        let elapsed = now_ns.saturating_sub(last);
        if elapsed > 0 {
            let refill = (elapsed as i64).saturating_mul(self.refill_ns);
            loop {
                let cur = self.tokens.load(Ordering::SeqCst);
                let next = (cur + refill).min(self.capacity);
                if self
                    .tokens
                    .compare_exchange(cur, next, Ordering::SeqCst, Ordering::SeqCst)
                    .is_ok()
                {
                    break;
                }
            }
        }

        loop {
            let cur = self.tokens.load(Ordering::SeqCst);
            if cur < NANO_TOKEN {
                return (cur / NANO_TOKEN, false);
            }
            if self
                .tokens
                .compare_exchange(cur, cur - NANO_TOKEN, Ordering::SeqCst, Ordering::SeqCst)
                .is_ok()
            {
                return ((cur - NANO_TOKEN) / NANO_TOKEN, true);
            }
        }
    }
}

pub struct RateLimiter {
    rps: f64,
    burst: i64,
    buckets: RwLock<HashMap<String, Arc<ClientBucket>>>,
}

impl RateLimiter {
    pub fn new(requests_per_second: f64, burst_size: Option<i64>) -> Self {
        let burst = burst_size.unwrap_or_else(|| requests_per_second.max(1.0) as i64);
        Self {
            rps: requests_per_second,
            burst,
            buckets: RwLock::new(HashMap::new()),
        }
    }

    pub fn allow(&self, key: &str) -> (i64, bool) {
        {
            let read = self.buckets.read().unwrap();
            if let Some(bucket) = read.get(key) {
                return bucket.allow();
            }
        }
        let mut write = self.buckets.write().unwrap();
        let bucket = write
            .entry(key.to_string())
            .or_insert_with(|| Arc::new(ClientBucket::new(self.rps, self.burst)))
            .clone();
        bucket.allow()
    }

    pub fn limit(&self) -> i64 {
        self.burst
    }

    pub fn rps(&self) -> f64 {
        self.rps
    }
}

const LUA_TOKEN_BUCKET: &str = r#"
local key = KEYS[1]
local rate = tonumber(ARGV[1])
local burst = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local tokens_key = key .. ":tokens"
local timestamp_key = key .. ":ts"

local last_tokens = tonumber(redis.call("get", tokens_key))
if last_tokens == nil then
    last_tokens = burst
end

local last_refreshed = tonumber(redis.call("get", timestamp_key))
if last_refreshed == nil then
    last_refreshed = 0
end

local delta = math.max(0, now - last_refreshed)
local filled_tokens = math.min(burst, last_tokens + (delta * rate))
local allowed = filled_tokens >= 1
local new_tokens = filled_tokens

if allowed then
    new_tokens = filled_tokens - 1
end

redis.call("setex", tokens_key, math.ceil(burst / rate), new_tokens)
redis.call("setex", timestamp_key, math.ceil(burst / rate), now)

return { new_tokens, allowed }
"#;

pub struct RedisRateLimiter {
    client: Client,
    rps: f64,
    burst: i64,
}

impl RedisRateLimiter {
    pub fn new(client: Client, rps: f64, burst_size: Option<i64>) -> Self {
        let burst = burst_size.unwrap_or_else(|| rps.max(1.0) as i64);
        Self { client, rps, burst }
    }

    pub async fn allow(&self, key: &str) -> redis::RedisResult<(i64, bool)> {
        let mut con = self.client.get_async_connection().await?;
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs_f64();

        let result: Vec<i64> = redis::Script::new(LUA_TOKEN_BUCKET)
            .key(format!("ratelimit:{}", key))
            .arg(self.rps)
            .arg(self.burst)
            .arg(now)
            .invoke_async(&mut con)
            .await?;

        if result.len() == 2 {
            Ok((result[0], result[1] == 1))
        } else {
            Err(redis::RedisError::from((
                redis::ErrorKind::TypeError,
                "Invalid response from Lua script",
            )))
        }
    }
}

